common/src/main/java/org/jboss/pnc/common/Configuration.java

```diff
@@ -33,19 +37,25 @@
      * @return Loaded configuration
      * @throws ConfigurationParseException Thrown if configuration file couldn't be loaded or parsed
      */
-    public T getModuleConfig(Class<T> moduleClass) throws ConfigurationParseException {
+    @SuppressWarnings("unchecked")
+    public <T extends AbstractModuleConfig> T getModuleConfig(Class<T> moduleClass) throws ConfigurationParseException {
```

Comment:
@slaskawi I got the example, but in our case we have just one instance and the field is initiated in constructor. So I think we do not need the volatile. Or am I wrong?

I expect it could look like this, right?:

    public <T extends AbstractModuleConfig> T getModuleConfig(Class<T> moduleClass) throws ConfigurationParseException {
        if(configCache.containsKey(moduleClass))
            return (T) configCache.get(moduleClass);
        
        synchronized(this) {
            if(configCache.containsKey(moduleClass))
                return (T) configCache.get(moduleClass);
            
            try (InputStream configStream = this.getConfigStream()) {
                log.info("Loading configuration for class: " + moduleClass);
                String configString = StringUtils.replaceEnv(IoUtils.readStreamAsString(configStream));
                
                T config = new ConfigurationJSONParser().parseJSONConfig(configString, moduleClass);
                configCache.put(moduleClass, config);
                return config;
            } catch (IOException e) {
                throw new ConfigurationParseException("Config could not be parsed", e);
            }
        }
    }